nixos/modules/config/networking.nix

```diff
@@ -188,11 +214,23 @@ in
 
         # /etc/hosts: Hostname-to-IP mappings.
         "hosts".text =
+          let oneToString = set : ip : ip + " " + concatStringsSep " " ( getAttr ip set );
+              allToString = set : concatMapStringsSep "\n" ( oneToString set ) ( attrNames set );
+              userLocalHosts = optionalString
+                ( builtins.hasAttr "127.0.0.1" cfg.hosts )
+                ( concatStringsSep " " ( remove "localhost" cfg.hosts."127.0.0.1" ));
+              userLocalHosts6 = optionalString
+                ( builtins.hasAttr "::1" cfg.hosts )
+                ( concatStringsSep " " ( remove "localhost" cfg.hosts."::1" ));
+              otherHosts = allToString ( removeAttrs cfg.hosts [ "127.0.0.1" "::1" ]);
+              maybeFQDN = optionalString ( cfg.fqdn != null ) cfg.fqdn;
```

Comment:
> we can still provide our own merge function instead of the default one.

This. Although I have no idea how to implement it.

> mapping canonical hostnames to IP addresses.

This is good too. Current implementation misses check that no hostname mapped to two different IP addresses, but this way it can be done trivially.

In my mind, it looks like this:
```Nix
{
  networking.hosts.aliases = {
    "foo.example.com" = "127.0.0.2";
    "bar.example.com" = "127.0.0.2";
  };
  networking.hosts.canonical = {
    "foobar.example.com" = "127.0.0.2";
  };
}
```
which will emit `127.0.0.2 foobar.example.com foo.example.com bar.example.com`.